indico/modules/search/client/js/components/SearchApp.jsx

```diff
@@ -106,55 +88,85 @@ SearchTypeMenuItem.defaultProps = {
-    setActiveMenuItem(index);
-  };
+function useQueryParams() {
+  const [query, _setQuery] = useState(window.location.search);
+  const queryObject = useMemo(() => Object.fromEntries(new URLSearchParams(query)), [query]);
+  const history = useHistory();
+
+  const setQuery = useCallback(
+    (type, name, reset) => {
+      const params = new URLSearchParams(reset ? undefined : query);
+      if (name !== undefined) {
+        params.set(type, name);
+      } else {
+        params.delete(type);
+      }
+      const _query = params.toString();
+      history.push({
+        pathname: window.location.pathname,
+        search: `?${_query}`,
+      });
+      _setQuery(_query);
+    },
+    [history, query]
+  );
 
-  const [categoryResults, setCategoryPage] = useSearch(searchCategoryURL(), query);
-  const [eventResults, setEventPage] = useSearch(searchEventURL(), query);
-  const [contributionResults, setContributionPage] = useSearch(searchContributionURL(), query);
-  const [fileResults, setFilePage] = useSearch(searchAttachmentURL(), query);
+  useEffect(() => {
+    return history.listen(location => _setQuery(location.search));
+  }, [history, query, _setQuery]);
+
+  return [queryObject, setQuery];
+}
+
+export default function SearchApp() {
+  const [query, setQuery] = useQueryParams();
+  const [activeMenuItem, setActiveMenuItem] = useState(undefined);
+  const {q, ...filters} = query;
+  const [categoryResults, setCategoryPage] = useSearch(searchURL({type: 'category'}), query);
+  const [eventResults, setEventPage] = useSearch(searchURL({type: 'event'}), query);
+  const [contributionResults, setContributionPage] = useSearch(
+    searchURL({type: 'contribution'}),
+    query
+  );
+  const [fileResults, setFilePage] = useSearch(searchURL({type: 'attachment'}), query);
   const searchMap = [
     ['Categories', categoryResults, setCategoryPage, Category],
     ['Events', eventResults, setEventPage, Event],
     ['Contributions', contributionResults, setContributionPage, Contribution],
     ['Materials', fileResults, setFilePage, File],
   ];
+  // Defaults to the first tab loading or with results
+  const menuItem =
+    activeMenuItem || Math.max(0, searchMap.findIndex(x => x[1].loading || x[1].total));
   // eslint-disable-next-line no-unused-vars
-  const [_, results, setPage, Component] = searchMap[activeMenuItem];
+  const [_, results, setPage, Component] = searchMap[menuItem];
   const isAnyLoading = searchMap.some(x => x[1].loading);
 
-  // On every search, the tab will be either the first loading or with results
-  const menuItem = searchMap.findIndex(x => x[1].loading || x[1].total > 0);
-  // TODO: we could make this initially uncontrolled, but if the user controls it, it no longer changes
-  useEffect(() => {
-    setActiveMenuItem(Math.max(0, menuItem));
-  }, [menuItem]);
-
-  const handleQuery = value => setQuery(value, 'pushIn');
+  const handleQuery = (value, type) => {
+    const _type = type || 'q';
```

Comment:
Not sure if it works, but in case it gets called without an argument for type (and not just a falsy one), you could just use a default value:

```suggestion
  const handleQuery = (value, type = 'q') => {
```

Reply:
Sure, neither the value should be strictly undefined or false, so this is fine too.